### PR TITLE
`project/create` and `project/delete` methods

### LINKED ISSUE #527: Implement Project Creation and Deletion
### Summary
The next key milestone for the project manager is the ability to create new projects and delete projects.

### Value
The IDE will be able to create and delete projects.

### Specification

- [ ] Implement the `project/create` message.
- [ ] Create a project in response to the message.
- [ ] Implement the `project/delete` message.
- [ ] Delete a project in response to the message.

### Acceptance Criteria & Test Cases

- The above specification has been completed.
- The listed features have been rigorously tested.

### PR DESCRIPTION
**Pull request by lolczak**
### Pull Request Description
Adds to the project manager the ability to create new projects and delete projects.

closes #527 

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: common/project-manager/src/main/scala/org/enso/projectmanager/service/MtlProjectValidator.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import cats.implicits._
+import org.enso.projectmanager.service.ValidationFailure.{
+  EmptyName,
+  NameContainsForbiddenCharacter
+}
+
+/**
+  * MTL implementation of the project validator.
+  */
+class MtlProjectValidator[F[_, _]](
```
Comment (kustosz) [inline common/project-manager/src/main/scala/org/enso/projectmanager/service/MtlProjectValidator.scala]: I don't like `Mtl` being there in the name – it's just the name of a library, not a concept. Maybe `MonadicProjectValidator`. Or `MonadErrorProjectValidator`?

File: common/project-manager/src/test/scala/org/enso/projectmanager/protocol/ProjectManagementApiSpec.scala [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+      client.expectJson(json"""
+          {
+            "jsonrpc" : "2.0",
+            "id" : 1,
+            "result" : {
+              "projectId" : $TestUUID
+            }
+          }
+          """)
+      packageFile shouldBe 'file
```
Comment (kustosz) [inline common/project-manager/src/test/scala/org/enso/projectmanager/protocol/ProjectManagementApiSpec.scala]: AFAIR this syntax is deprecated now, in lieu of `Symbol("file")`

File: doc/design/engine/engine-services.md [hunk 3]
```diff
@@ -729,7 +728,10 @@ interface ProjectCreateRequest {
 ```
 
 ##### Errors
-TBC
+- [`ProjectNameValidationError`](#projectnamevalidationerror) to signal validation failures.
```
Comment (kustosz) [inline doc/design/engine/engine-services.md]: Wrap this file to 80 chars.

File: common/project-manager/src/main/scala/org/enso/projectmanager/control/core/CovariantFlatMap.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package org.enso.projectmanager.control.core
+
+import zio._
+
+/**
+  * A class for covariant bifunctors used to chain computations.
```
[the remark this answers is not in the document]
Comment (lolczak) [reply common/project-manager/src/main/scala/org/enso/projectmanager/control/core/CovariantFlatMap.scala]: fixed

File: common/project-manager/src/main/scala/org/enso/projectmanager/control/core/CovariantFlatMap.scala [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  def pure[A](value: A): F[Nothing, A]
+
+  /**
+    * Chains `fa` with `f` arrow.
+    *
+    * @param fa an effectful computation
+    * @param f dependent effectful function
+    * @return a composition of dependent effectful functions
+    */
+  def flatMap[E1, E2 >: E1, A, B](fa: F[E1, A])(f: A => F[E2, B]): F[E2, B]
```
[the remark this answers is not in the document]
Comment (lolczak) [reply common/project-manager/src/main/scala/org/enso/projectmanager/control/core/CovariantFlatMap.scala]: It's needed by Scala compiler to detect the upper bound of errors. F context is covariant on first type parameter.

File: common/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/file/SynchronizedFileStorage.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+  * ZIO implementation of [[FileStorage]]. It uses circe [[Encoder]] and
+  * [[Decoder]] to encode/decode objects.
+  *
+  * @param path a path to a file that stores serialized object
+  * @param fileSystem a filesystem algebra
+  * @tparam A a datatype to store
+  */
+class SynchronizedFileStorage[A: Encoder: Decoder: Default, F[+_, +_]: Sync: ErrorChannel: CovariantFlatMap](
```
[the remark this answers is not in the document]
Comment (lolczak) [reply common/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/file/SynchronizedFileStorage.scala]: scalafmt doesn't break type constraints.

File: common/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/repository/ProjectFileRepository.scala [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  ): F[ProjectRepositoryFailure, Unit] =
+    indexStorage
+      .modify { index =>
+        val maybeProject = index.findUserProject(projectId)
+        index.removeUserProject(projectId) -> maybeProject
+      }
+      .mapError(_.fold(convertFileStorageFailure))
+      .flatMap {
+        case None =>
+          ErrorChannel[F].fail[ProjectRepositoryFailure](ProjectNotFoundInIndex)
```
[the remark this answers is not in the document]
Comment (lolczak) [reply common/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/repository/ProjectFileRepository.scala]: Actually, the scala compiler infers it automatically, but the Intellij compiler does not. Fixed.

File: common/project-manager/src/main/scala/org/enso/projectmanager/service/MtlProjectValidator.scala (hunk 1 above)
Comment (lolczak) [reply common/project-manager/src/main/scala/org/enso/projectmanager/service/MtlProjectValidator.scala]: Heh, in Scala it is a concept like [here](https://typelevel.org/blog/2018/11/28/http4s-error-handling-mtl-2.html). But, it is a good point. Fixed.

File: common/project-manager/src/test/scala/org/enso/projectmanager/protocol/ProjectManagementApiSpec.scala (hunk 2 above)
Comment (lolczak) [reply common/project-manager/src/test/scala/org/enso/projectmanager/protocol/ProjectManagementApiSpec.scala]: fixed

File: doc/design/engine/engine-services.md (hunk 3 above)
Comment (lolczak) [reply doc/design/engine/engine-services.md]: fixed

File: common/project-manager/src/main/scala/org/enso/projectmanager/service/MtlProjectValidator.scala (hunk 1 above)
Comment (kustosz) [reply common/project-manager/src/main/scala/org/enso/projectmanager/service/MtlProjectValidator.scala]: Uggggh, that's... something. Anyway, let's stick to calling this stuff monad transformers and MTL being just a particular implementation.

**Pull request merged**